include/clang/StaticAnalyzer/Core/BugReporter/InternalInfo.h

```diff
@@ -0,0 +1,99 @@
+//===- InternalInfo.h -------------------------------------------*- C++ -*-===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file defines the InternalInfo interfaces.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_STATICANALYZER_CORE_BUGREPORTER_INTERNALINFO_H
+#define LLVM_CLANG_STATICANALYZER_CORE_BUGREPORTER_INTERNALINFO_H
+
+#include "clang/StaticAnalyzer/Core/PathSensitive/ProgramState.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/Support/raw_ostream.h"
+#include <map>
+
+namespace clang {
+namespace ento {
+
+class InternalInfo {
+public:
+  enum Kind { None, Constraint, State};
+
+private:
+  const Kind kind;
+  static constexpr Kind getStaticKind() { return None; }
+
+public:
+  InternalInfo(Kind kind) : kind(kind) {}
+  virtual ~InternalInfo() = default;
+
+  Kind getKind() const { return kind; }
+
+
+  static bool isKind(const InternalInfo& Info) {
+    return Info.getKind() == getStaticKind();
+  }
+
+  template <typename T>
+  T* getAs() {
```

Comment:
I think LLVM syntax requires this to be written as `T *getAs()` (even though it's really ugly).